lib/std/child_process.zig

```diff
@@ -257,6 +259,55 @@ pub const ChildProcess = struct {
         }
     }
 
+    fn collectOutputWindows(child: *ChildProcess, outs: [2]*std.ArrayList(u8), max_output_bytes: usize) !void {
+        const bump_amt = 512;
+        const handles = [_]windows.HANDLE{ child.stdout.?.handle, child.stderr.?.handle };
+
+        for ([_]u1{ 0, 1 }) |i| {
+            try outs[i].ensureCapacity(bump_amt);
+            const buf = outs[i].unusedCapacitySlice();
+            child.overlapped[i] = mem.zeroes(windows.OVERLAPPED);
+            _ = windows.kernel32.ReadFile(handles[i], buf.ptr, @intCast(u32, buf.len), null, &child.overlapped[i]);
+        }
+
+        var wait_objects = handles;
+        var wait_object_count: u2 = wait_objects.len;
+
+        while (true) {
+            const status = windows.kernel32.WaitForMultipleObjects(wait_object_count, &wait_objects, 0, windows.INFINITE);
+            if (status == windows.WAIT_FAILED) {
+                switch (windows.kernel32.GetLastError()) {
+                    else => |err| return windows.unexpectedError(err),
+                }
+            }
+            if (status < windows.WAIT_OBJECT_0 or status > windows.WAIT_OBJECT_0 + wait_object_count - 1)
+                unreachable;
+
+            const wait_idx = status - windows.WAIT_OBJECT_0;
+            const i = @as(u1, if (wait_objects[wait_idx] == handles[0]) 0 else 1);
+            var read_bytes: u32 = undefined;
+            const overlapped_result = windows.kernel32.GetOverlappedResult(handles[i], &child.overlapped[i], &read_bytes, 0);
+            if (overlapped_result == 0) switch (windows.kernel32.GetLastError()) {
+                .BROKEN_PIPE => {
+                    wait_object_count -= 1;
+                    if (wait_object_count == 0)
+                        break;
+                    if (wait_idx == 0)
+                        wait_objects[0] = wait_objects[1];
+                    continue;
+                },
+                else => |err| return windows.unexpectedError(err),
+            };
+
+            outs[i].items.len += read_bytes;
+            const new_capacity = std.math.min(outs[i].items.len + bump_amt, max_output_bytes);
+            try outs[i].ensureCapacity(new_capacity);
+            const buf = outs[i].unusedCapacitySlice();
+            if (buf.len == 0) return if (i == 0) error.StdoutStreamTooLong else error.StderrStreamTooLong;
+            _ = windows.kernel32.ReadFile(handles[i], buf.ptr, @intCast(u32, buf.len), null, &child.overlapped[i]);
```

Comment:
This `@intCast` is fishy, can `buf.len` become greater than `maxInt(u32)` and if so how do we handle this case?

Reply:
sure, we can and it with `0xFFFFFFFF`

Reply:
Can you detail how it's possible for the length to become this large?

You don't want to truncate the result, if anything you want something like `math.cast(u32, Len) catch maxInt(u32)`.

Reply:
> You don't want to truncate the result, if anything you want something like `math.cast(u32, Len) catch maxInt(u32)`.

why?

Reply:
Because you want to read as much data as possible rather than the possibly smlal remainder that you obtain after truncating the length.
0xffffffff is much closer to 0x100000000 than 0.

Reply:
I'm sorry I'm not understanding you here.  In my brain, `@intCast(u32, x & 0xFFFFFFFF)` is functionality equivalent to `match.cast(u32, x) catch maxInt(u32)`.

Reply:
Oh nevermind I just got it! sorry :)